### PR TITLE
Decrease reliance on curbuf in BUFEMPTY and `undo.c`

### PR DESCRIPTION
**Pull request by vigoux**
This avoids using curbuf in the `BUFEMPTY` macro.

This is my take on the effort towards using less `curbuf` and `curwin` (and started by @teto).

Should I add more commits like this one in here ?

### CODE REVIEW

**Review by teto — COMMENTED** [review]
I approve of the idea. I would prioritize "un-curbufying" cases that remove `save_curbuf` from the code.

File: src/nvim/macros.h [hunk 1]
```diff
@@ -35,8 +35,8 @@
 #define LINEEMPTY(p) (*ml_get(p) == NUL)
 
 /// BUFEMPTY() - return TRUE if the current buffer is empty
-#define BUFEMPTY() (curbuf->b_ml.ml_line_count == 1 && *ml_get((linenr_T)1) == \
-                    NUL)
+#define BUFEMPTY(buf) ((buf)->b_ml.ml_line_count == 1 \
```
Comment (bfredl) [inline src/nvim/macros.h:38]: if you feel like refactor I think this should be `static inline bool buf_empty(buf_T *buf)` in `buffer.h` (but if not we can do that later).

Comment (vigoux) [reply src/nvim/macros.h:38]: Sounds good to me then

File: src/nvim/macros.h [hunk 2]
```diff
@@ -35,8 +35,8 @@
 #define LINEEMPTY(p) (*ml_get(p) == NUL)
 
 /// BUFEMPTY() - return TRUE if the current buffer is empty
```
Comment (simonweil) [inline src/nvim/macros.h:37]: The comment is now outdated

Comment (vigoux) [reply src/nvim/macros.h:37]: Agreed, we might even remove the bufempty macro to move it as a function, thus I'll address that later on.

**Comment by vigoux** [comment]
I have been working on remove a use of `save_curbuf` in `undo`, but this is trickier then expected, thus causing a delay here. I'll try finishing that today.

**Comment by vigoux** [comment]
So, coming back to this PR, it seems it should be pretty much ready. I rebased, just to get it rolling as maybe part of 0.5.1 ?

**Comment by bfredl** [comment]
0.5.1 is for bug fixes tho, I wouldn't include this unless some bug (or QoL feature) depends on it directly.

**Comment by vigoux** [comment]
Then that'd be part of 0.6, so we don't push it yet?

**Comment by clason** [comment]
Current master _is_ (what will be) 0.6.0; patch versions like 0.5.1 are cut from the `release-0.5` branch. So this is OK to merge; it just won't be backported to that branch.

**Comment by vigoux** [comment]
@teto @bfredl i think this is pretty much ready.

I think I'll do other passes doing the same thing now that `curbuf_lock` is removed, as this should enable quite a lot of things.

File: src/nvim/buffer_defs.h [hunk 3]
```diff
@@ -525,6 +525,7 @@ struct file_buffer {
   int b_flags;                  // various BF_ flags
   int b_locked;                 // Buffer is being closed or referenced, don't
                                 // let autocommands wipe it out.
+  int b_ro_locked;              // Replacement for curbuf_lock
```
Comment (bfredl) [inline src/nvim/buffer_defs.h:528]: meaningless as `curbuf_lock` no longer exists, instead something like the removed comment there  `/// Non-zero when the buffer can't be changed.  Used for FileChangedRO.`

File: src/nvim/ex_getln.c [hunk 4]
```diff
@@ -2412,13 +2412,11 @@ char_u * get_text_locked_msg(void) {
   }
 }
 
-/*
- * Check if "curbuf_lock" or "allbuf_lock" is set and return TRUE when it is
- * and give an error message.
- */
+/// Check if "curbuf_lock" or "allbuf_lock" is set and return TRUE when it is
```
Comment (bfredl) [inline src/nvim/ex_getln.c:2415]: `curbuf->b_ro_locked`

**Comment by bfredl** [comment]
https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim when merging could add `BUFEMPTY()` to the list of functions here

File: src/nvim/ex_getln.c (hunk 4 above)
Comment (vigoux) [reply src/nvim/ex_getln.c]: Done

File: src/nvim/buffer_defs.h (hunk 3 above)
Comment (vigoux) [reply src/nvim/buffer_defs.h]: Done

File: src/nvim/macros.h (hunk 2 above)
Comment (vigoux) [reply src/nvim/macros.h]: Done

File: src/nvim/macros.h (hunk 1 above)
Comment (vigoux) [reply src/nvim/macros.h]: Done

File: src/nvim/undo.c [hunk 5]
```diff
@@ -3146,11 +3176,13 @@ u_header_T *u_force_get_undo_header(buf_T *buf)
     buf_T *save_curbuf = curbuf;
     curbuf = buf;
     // Args are tricky: this means replace empty range by empty range..
-    u_savecommon(0, 1, 1, true);
+    u_savecommon(curbuf, 0, 1, 1, true);
+    curbuf = save_curbuf;
```
Comment (bfredl) [inline src/nvim/undo.c:3180]: this doesn't look right. perhaps this `save_curbuf` declaration can be eliminated now?

File: src/nvim/testdir/otherName [hunk 6]
```diff
@@ -0,0 +1,2 @@
+some
```
Comment (bfredl) [inline src/nvim/testdir/otherName:1]: is this file used?

**Comment by bfredl** [comment]
LTGM after comments :)

**Comment by vigoux** [comment]
Should be all good for real now

**Pull request merged**